include/deal.II/base/index_set.h

```diff
@@ -497,6 +689,246 @@ IndexSet complete_index_set (const unsigned int N)
+  : accessor(idxset, idxset.ranges.size())
+{}
+
+inline
+IndexSet::IntervalIterator &
+IndexSet::IntervalIterator::operator++ ()
+{
+  accessor.advance();
+  return *this;
+}
+
+inline
+IndexSet::IntervalIterator
+IndexSet::IntervalIterator::operator++ (int)
+{
+  const IndexSet::IntervalIterator iter = *this;
+  accessor.advance ();
+  return iter;
+}
+
+inline
+const IndexSet::IntervalAccessor &
+IndexSet::IntervalIterator::operator* () const
+{
+  return accessor;
+}
+
+inline
+const IndexSet::IntervalAccessor *
+IndexSet::IntervalIterator::operator-> () const
+{
+  return &accessor;
+}
+
+inline
+bool IndexSet::IntervalIterator::operator == (const IndexSet::IntervalIterator &other) const
+{
+  return accessor == other.accessor;
+}
+
+inline
+bool IndexSet::IntervalIterator::operator != (const IndexSet::IntervalIterator &other) const
+{
+  return !(*this == other);
+}
+
+inline
+bool IndexSet::IntervalIterator::operator < (const IndexSet::IntervalIterator &other) const
+{
+  return accessor < other.accessor;
+}
+
+inline
+std::ptrdiff_t IndexSet::IntervalIterator::operator - (const IndexSet::IntervalIterator &other) const
+{
+  Assert (accessor.index_set == other.accessor.index_set, ExcInternalError());
+
+  return static_cast<std::ptrdiff_t>(accessor.range_idx) - static_cast<std::ptrdiff_t>(other.accessor.range_idx);
+}
+
+
+/* ElementIterator */
+
+inline
+bool IndexSet::ElementIterator::is_valid() const
+{
+  return range_idx < index_set.ranges.size() && idx<index_set.ranges[range_idx].end;
+}
+
+inline
+IndexSet::ElementIterator::ElementIterator(const IndexSet &idxset, IndexSet::size_type range_idx, IndexSet::size_type index)
+  : index_set(idxset), range_idx(range_idx), idx(index)
+{
+  Assert(range_idx < index_set.ranges.size(), ExcMessage(""));
+  Assert(idx >= index_set.ranges[range_idx].begin, ExcMessage(""));
+  Assert(idx <= index_set.ranges[range_idx].end, ExcMessage(""));
+}
+
+inline
+IndexSet::ElementIterator::ElementIterator(const IndexSet &idxset)
+  : index_set(idxset), range_idx(0), idx(0)
+{
+  if (index_set.ranges.size()>0)
+    {
+      range_idx = index_set.ranges.size()-1;
+      idx = index_set.ranges[range_idx].end;
+    }
+}
+
+inline
+const IndexSet::size_type &
+IndexSet::ElementIterator::operator* () const
+{
+  Assert(range_idx < index_set.ranges.size(), ExcMessage(""));
+  return idx;
+}
+
+inline
+bool IndexSet::ElementIterator::operator == (const IndexSet::ElementIterator &other) const
+{
+  Assert (index_set == other.index_set, ExcInternalError());
+  return range_idx == other.range_idx && idx==other.idx;
+}
+
+inline
+void IndexSet::ElementIterator::advance ()
+{
+  Assert(is_valid(), ExcMessage(""));
+  if (idx < index_set.ranges[range_idx].end)
+    ++idx;
+  if (idx == index_set.ranges[range_idx].end)
+    {
+      if (range_idx < index_set.ranges.size()-1)
+        {
+          ++range_idx;
+          idx = index_set.ranges[range_idx].begin;
+        }
```

Comment:
what happens in the `else` case? My experience with the sparsity pattern iterators was that rather than leaving the iterator in a sort-of-invalid state (`range_idx` is still valid, `idx` isn't any more), it's worthwhile explicitly setting the whole thing to an invalid iterator, e.g., as in 
```
  else
    // we fell off the end -- explicitly set iterator to the end-iterator
    *this = ElementIterator(index_set);
```